### PR TITLE
Implement RequestedRangeNotSatisfiedException for Azure and GCP

### PR DESCRIPTION
**Pull request by volodk85**
Make Azure and GCP BlobContainers throw `RequestedRangeNotSatisfiedException` if requested range (position & length) is beyond actual blob length.

ES-8341

### CODE REVIEW

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java
Comment (kingherc) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java:77]: This looks in line with Azure [documentation](https://learn.microsoft.com/en-us/rest/api/storageservices/specifying-the-range-header-for-blob-service-operations#format-1-bytesstartbyte-) which talks about 416 error code.

File: modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java
Comment (kingherc) [inline modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java:78]: Hmm, doesn't seem like this test fits in the "Retries" tests for Azure.

Should we better add support for returning `RestStatus.REQUESTED_RANGE_NOT_SATISFIED` in `AzureHttpHandler.java`, similar to how it's done in `S3HttpHandler`, and then add a test like `testReadFromPositionLargerThanBlobLength` (which simply does an invalid request) to `AzureStorageCleanupThirdPartyTests`?

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (kingherc) [inline modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:157]: Can we add a test like `testReadFromPositionLargerThanBlobLength` in `GoogleCloudStorageThirdPartyTests`? This means you'll need to also modify `GoogleCloudStorageHttpHandler` to send `RestStatus.REQUESTED_RANGE_NOT_SATISFIED` for invalid range offsets.

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (kingherc) [inline modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:163]: This seems in line with GCP documentation [here](https://cloud.google.com/storage/docs/json_api/v1/status-codes#416_Requested_Range_Not_Satisfiable).

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (kingherc) [inline modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:167]: Should here (and below line) be `Math.addExact(start, currentOffset)` instead of just `start`?

**Review by kingherc — COMMENTED** [review]
Goes to the right direction! Some first comments.

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (volodk85) [reply modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java]: thank you @kingherc

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (volodk85) [reply modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:157]: Thank you for a hint. Since all 3 cloud blob containers handle 416 error, there is a way to generalize the test for all 3 instances of 3rd part tests

File: modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java
Comment (volodk85) [reply modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java]: Agree. This adds to testing code clarity

**Review by kingherc — APPROVED** [review]
LGTM! Feel free to get also @tlrx 's approval before merging, as he was quite involved with introducing RequestedRangeNotSatisfiedException.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java
Comment (tlrx) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java:77]: Is there a reason we're rethrowing as a `RequestedRangeNotSatisfiedException` only if length is not null?

According to the Azure document, it is possible to only set the startByte (ie, offset) without a length and if that one is invalid, the service returns a 416. That's also how the AzureHttpHandler seems to be implemented.

I think we should just rethrow a `RequestedRangeNotSatisfiedException` here, possibly passing Long.MAX_VALUE to the RequestedRangeNotSatisfiedException.

File: test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java
Comment (tlrx) [inline test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:335]: nit: can we check explicitely RequestedRangeNotSatisfiedException#getPosition and getLength?

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (tlrx) [inline modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:90]: nit: I think that the previous constructor could use this one using `this(client, () -> getStorage(client), ....)` if we were passing a Function<Storage, Storage> function?

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java
Comment (tlrx) [inline modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java:56]: nit: unnecessary throws

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java
Comment (tlrx) [inline modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java:51]: position and length are always non null here, should we randomize this?

**Review by tlrx — COMMENTED** [review]
Looks good, I left some comments. This could have been 2 separate PRs :)

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java
Comment (volodk85) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java:77]: not null check is needed to avoid NPE during constructing RequestedRangeNotSatisfiedException (it takes primitive `long` argument), also not null `length` is somewhat meaningful in the context of RequestedRangeNotSatisfiedException since this exception is thrown only if length param is of the range

I think if we pass Long.MAX_VALUE we loose piece of real info which lead to exception.

File: modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
Comment (volodk85) [reply modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java:90]: Not a nit. It makes code much cleaner! Thank you

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java
Comment (volodk85) [reply modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java:51]: position and length are already randomized just few lines above.
(If I misread your comment and you meant something else, please rephrase)

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java
Comment (tlrx) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java:77]: > I think if we pass Long.MAX_VALUE we loose piece of context (real value) which led to exception.

Sorry, I was suggesting to rethrow `RequestedRangeNotSatisfiedException` also when `length` is null. 

In the code today we allow passing a `null` length down the `AzureBlobStore.getInputStream()` method which can throw a `RequestedRangeNotSatisfiedException`, but if length is null we don't rethrow it. I think we should, and to do that we need to build  a `RequestedRangeNotSatisfiedException` that requires a `long` length value: if it's null we can pass Long.MAX_VALUE or `-1`.

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java
Comment (tlrx) [reply modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java:51]: Sorry if I wasn't clear.

We're only testing reads with a position and a length, if we were passing null we could also test the `GoogleCloudStorageRetryingInputStream(Storage client, BlobId blobId)` default constructor I think.

Also, we're not checking the bytes that are effectively read, I think we should.

Comment (volodk85) [reply modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java]: Default constructor is not handy for unit testing because of its static init inside. I added one more (just for testing) which allows to plug mock object and keep semantics of defaul constructor `start: 0, end: Long.MAX_VALUE - 1`

Added a check for return bytes against requested

**Review by tlrx — APPROVED** [review]
LGTM, thanks for the extra iterations

**Pull request merged**
